apps/src/lib/levelbuilder/code-docs-editor/ProgrammingMethodEditor.jsx

```diff
@@ -0,0 +1,207 @@
+import PropTypes from 'prop-types';
+import React, {useState} from 'react';
+import OrderableList from './OrderableList';
+import ExampleEditor from './ExampleEditor';
+import ParameterEditor from './ParameterEditor';
+import TextareaWithMarkdownPreview from '@cdo/apps/lib/levelbuilder/TextareaWithMarkdownPreview';
+import CollapsibleEditorSection from '@cdo/apps/lib/levelbuilder/CollapsibleEditorSection';
+import HelpTip from '@cdo/apps/lib/ui/HelpTip';
+import SaveBar from '@cdo/apps/lib/levelbuilder/SaveBar';
+import {createUuid, navigateToHref} from '@cdo/apps/utils';
+import $ from 'jquery';
+import color from '@cdo/apps/util/color';
+
+function useProgrammingMethod(initialProgrammingMethod) {
+  const [programmingMethod, setProgrammingMethod] = useState(
+    initialProgrammingMethod
+  );
+
+  function updateProgrammingMethod(key, value) {
+    setProgrammingMethod({...programmingMethod, [key]: value});
```

Comment:
hmm I see what you're saying but I'm not sure the best way around it. `updateProgrammingMethod` is kind of a helper to prevent accidentally overwriting parts of the state and it came first. Then I found in a number of cases I needed the ability to reset the state, which is why that update function is also returned (so yeah, it is used in a number of places).

I could obviously just get rid of this hook and have everything use `setProgrammingMethod` but having `updateProgrammingMethod` has felt clearer to me most of the time and it seems like it would be less error prone to write `updateProgrammingMethod(name, e.target.value)` than `setProgrammingMethod({...programmingMethod, name: e.target.value)`.

Reply:
Ah, I see.  Maybe just rename `updateProgrammingMethod` to `setProgrammingMethodProperty` or something like that?

Reply:
Sure, I can do that